libnavui-maps/src/main/java/com/mapbox/navigation/ui/maps/puck/LocationPuckOptions.kt

```diff
@@ -0,0 +1,175 @@
+        result = 31 * result + arrivalPuck.hashCode()
+        return result
+    }
+
+    /**
+     * Regenerate whenever a change is made
+     */
+    override fun toString(): String {
+        return "LocationPuckOptions(" +
+            "context=$context, " +
+            "freeDrivePuck=$freeDrivePuck, " +
+            "destinationPreviewPuck=$destinationPreviewPuck, " +
+            "routePreviewPuck=$routePreviewPuck, " +
+            "activeNavigationPuck=$activeNavigationPuck, " +
+            "arrivalPuck=$arrivalPuck" +
+            ")"
+    }
+
+    /**
+     * Builder of [LocationPuckOptions]
+     */
+    class Builder(private val context: Context) {
+        private var freeDrivePuck = regularPuck()
+        private var destinationPreviewPuck = regularPuck()
+        private var routePreviewPuck = regularPuck()
+        private var activeNavigationPuck = navigationPuck()
+        private var arrivalPuck = navigationPuck()
+
+        /**
+         * Apply [LocationPuck2D] or [LocationPuck3D] to location puck in free drive state
+         * @param freeDrivePuck [LocationPuck] to be used in free drive state
+         */
+        fun freeDrivePuck(freeDrivePuck: LocationPuck): Builder = apply {
+            this.freeDrivePuck = freeDrivePuck
+        }
+
+        /**
+         * Apply [LocationPuck2D] or [LocationPuck3D] to location puck in destination preview state
+         * @param destinationPreviewPuck [LocationPuck] to be used in destination preview state
+         */
+        fun destinationPreviewPuck(destinationPreviewPuck: LocationPuck): Builder = apply {
+            this.destinationPreviewPuck = destinationPreviewPuck
+        }
+
+        /**
+         * Apply [LocationPuck2D] or [LocationPuck3D] to location puck in route preview state
+         * @param routePreviewPuck [LocationPuck] to be used in route preview state
+         */
+        fun routePreviewPuck(routePreviewPuck: LocationPuck): Builder = apply {
+            this.routePreviewPuck = routePreviewPuck
+        }
+
+        /**
+         * Apply [LocationPuck2D] or [LocationPuck3D] to location puck in active navigation state
+         * @param activeNavigationPuck [LocationPuck] to be used in active navigation state
+         */
+        fun activeNavigationPuck(activeNavigationPuck: LocationPuck): Builder = apply {
+            this.activeNavigationPuck = activeNavigationPuck
+        }
+
+        /**
+         * Apply [LocationPuck2D] or [LocationPuck3D] to location puck in arrival state
+         * @param arrivalPuck [LocationPuck] to be used in arrival state
+         */
+        fun arrivalPuck(arrivalPuck: LocationPuck): Builder = apply {
+            this.arrivalPuck = arrivalPuck
+        }
+
+        private fun navigationPuck(): LocationPuck = LocationPuck2D(
+            bearingImage = ContextCompat.getDrawable(
+                context,
+                R.drawable.mapbox_navigation_puck_icon,
+            )
+        )
+
```

Comment:
Let's add `defaultPuck(LocationPuck)` (or something similar) that can be used to simplify setting custom puck for all states.

```kotlin
// instead of writing this
val options = LocationPuckOptions.Builder(context)
    .freeDrivePuck(myPuck)
    .destinationPreviewPuck(myPuck)
    .routePreviewPuck(myPuck)
    .activeNavigationPuck(myPuck)
    .arrivalPuck(myPuck)
    .build()

// devs could write just this
val options = LocationPuckOptions.Builder(context)
    .defaultPuck(myPuck)
    .build()

// or this
val options = LocationPuckOptions.Builder(context, defaultPuck = myPuck)
    .build()
```

Thoughts?

Reply:
Implemented.